Allow setting mac settings when creating devices in the Console

bafonins opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

References https://github.com/TheThingsNetwork/lorawan-stack/issues/2047

<details>
<summary>Screenshots</summary>
Device wizard network settings step:

OTAA:

<img width="834" alt="Screenshot 2020-11-03 at 17 10 37" src="https://user-images.githubusercontent.com/16374166/98003757-c8ec7780-1df7-11eb-8e2b-67a03c8e961d.png">

OTAA class B:

<img width="834" alt="Screenshot 2020-11-03 at 17 10 45" src="https://user-images.githubusercontent.com/16374166/98003762-cb4ed180-1df7-11eb-9cd2-23c1b29aeecc.png">


ABP:

<img width="834" alt="Screenshot 2020-11-03 at 17 11 06" src="https://user-images.githubusercontent.com/16374166/98003803-d6096680-1df7-11eb-8fe2-7100fff80fa4.png">

ABP class B:

<img width="834" alt="Screenshot 2020-11-03 at 17 11 13" src="https://user-images.githubusercontent.com/16374166/98003810-d86bc080-1df7-11eb-9a15-f777b3002de0.png">


Multicast:

<img width="834" alt="Screenshot 2020-11-03 at 17 11 33" src="https://user-images.githubusercontent.com/16374166/98003898-f76a5280-1df7-11eb-913a-65014c95ec62.png">

Multicast class B:

<img width="834" alt="Screenshot 2020-11-03 at 17 12 01" src="https://user-images.githubusercontent.com/16374166/98003903-f9ccac80-1df7-11eb-9b56-1af7d78dfac2.png">


General settings example:
<img width="771" alt="Screenshot 2020-11-03 at 17 15 01" src="https://user-images.githubusercontent.com/16374166/98004032-1c5ec580-1df8-11eb-8ced-6851bbaa59de.png">



</details>

#### Changes
<!-- What are the changes made in this pull request? -->

- Adjust the `<KeyValueMap />` component to allow using indices as keys
- Extend the network settings step in the device wizard to allow setting mac settings
- Various adjustments for the wizard component


#### Testing

<!-- How did you verify that this change works? -->

Manual testing for various device types + e2e tests for end device creation.


##### Regressions

<!-- Please indicate features that this change could affect and how that was tested. -->

Device creation or just the network settings step in the device wizard.


#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

Commit:
console: Add device MAC settings page

Commit:
console: Extend key-value-map component

allow using array indices as keys

Commit:
console: Add supports_class_b to device general settings

Commit:
console: Add supports_class_b to device wizard

Commit:
console,oauth: Fix form section style

Commit:
console: Add mac settings section

Commit:
console: Add mac settings to network settings form

Commit:
console,oauth: Adjust select prop-types

Commit:
doc: Update changelog

kschiffer commented on pkg/webui/console/components/mac-settings-section/index.js:
```diff
@@ -0,0 +1,194 @@
+  rx2FrequencyDescription: 'Frequency for RX2 (Hz)',
+  frequencyPlaceholder: 'e.g. 869525000 for 869,525 MHz',
+  macSettings: 'MAC settings',
+  updateSuccess: 'The MAC settings have been updated successfully',
```
```suggestion
  updateSuccess: 'The MAC settings updated',
```

kschiffer commented on pkg/webui/components/key-value-map/entry.js:
```diff
@@ -84,12 +89,16 @@ class Entry extends React.Component {
+  value: PropTypes.shape({
+    key: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+    value: PropTypes.any,
+  }).isRequired,
```
```suggestion
  }),
```

In your implementation, the value can be `undefined`.

kschiffer commented on pkg/webui/console/components/mac-settings-section/index.js:
```diff
@@ -0,0 +1,194 @@
+  return (
+    <Form.CollapseSection
+      id="mac-settings"
+      title={m.macSettings}
```
```suggestion
      title={sharedMessages.advancedSettings}
```

kschiffer commented on pkg/webui/console/components/mac-settings-section/index.js:
```diff
@@ -0,0 +1,194 @@
+      title={m.macSettings}
+      initiallyCollapsed={initiallyCollapsed}
+    >
+      {isABP && (
```
```suggestion
      {isABP && (
      <Form.SubTitle title={m.macSettings} />
```

kschiffer commented on pkg/webui/console/components/mac-settings-section/index.js:
```diff
@@ -0,0 +1,194 @@
+        title={m.rx2DataRateIndexTitle}
+        description={m.rx2DataDateIndexDescription}
+        name="mac_settings.rx2_data_rate_index.value"
+        component={Select}
```
Why not `type="number"` and `component={Input}` ?

kschiffer commented on pkg/webui/console/components/mac-settings-section/index.js:
```diff
@@ -0,0 +1,194 @@
+        <Form.Field
+          title={m.rx1DataRateOffsetTitle}
+          name="mac_settings.rx1_data_rate_offset"
+          component={Select}
```
Why not `type="number"` and `component={Input}` ?

kschiffer commented on pkg/webui/console/components/mac-settings-section/index.js:
```diff
@@ -0,0 +1,194 @@
+        <Form.Field
+          title={m.rx1DelayTitle}
+          description={m.rx1DelayDescription}
+          name="mac_settings.rx1_delay.value"
```
Why not `type="number"` and `component={Input}` ?

kschiffer commented on pkg/webui/console/components/mac-settings-section/index.js:
```diff
@@ -0,0 +1,194 @@
+        description={m.rx2FrequencyDescription}
+        placeholder={m.frequencyPlaceholder}
+        name="mac_settings.rx2_frequency"
+        component={Input}
```
Can we use `encode`/ `decode` here so the user can use the common floating point notation?

bafonins replied:
Not sure what you mean

kschiffer replied:
`869,525` MHz frequency needs to be entered as `869525000` Hz. I suspect that it's more intuitive for the user to be able to enter the value as MHz. This can be done via the `encode` / `decode` props.

kschiffer replied:
Actually, even better would be to use the `<UnitInput />` component here, with a default on `MHz` and `Hz` second option.

kschiffer commented on pkg/webui/console/components/mac-settings-section/index.js:
```diff
@@ -0,0 +1,194 @@
+            title={m.pingSlotPeriodicityTitle}
+            description={m.pingSlotPeriodicityDescription}
+            name="mac_settings.ping_slot_periodicity.value"
+            component={Select}
```
Why `Select`?

bafonins replied:
What else?

kschiffer replied:
I'm not familiar with this setting but in depth, but if the ping slot periodicity allows for a range of numbers, we should rather use the `type="number"` input.

bafonins replied:
`ping_slot_periodicity` enum values can be mapped to a numeric value (see https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.9/api/api.md#enum-pingslotperiod). but is it really more convenient for the user to specify `0` to set the period to 1s, instead of just choosing `every second` in the select?

Also related https://github.com/TheThingsNetwork/lorawan-stack/issues/2258

kschiffer replied:
Ok yeah, I overlooked how the value is composed and `<Select />` is indeed the way to go here.

kschiffer commented on pkg/webui/console/components/mac-settings-section/index.js:
```diff
@@ -0,0 +1,194 @@
+          title={m.rx1DataRateOffsetTitle}
+          name="mac_settings.rx1_data_rate_offset"
+          component={Select}
+          options={dataRateOffsetOptions}
```
Can we add a description here too?

kschiffer commented on pkg/webui/console/components/mac-settings-section/mac-settings-section.js:
```diff
@@ -0,0 +1,199 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
```
I guess this file is a duplicate?

kschiffer commented on pkg/webui/console/components/mac-settings-section/validation-schema.js:
```diff
@@ -0,0 +1,124 @@
+    }),
+    rx2_frequency: Yup.number().min(100000, {
+      ...sharedMessages.validateNumberGte,
+      values: { value: 100000 },
```
I guess this is a remainder from before we introduced  `yup.passValues`.

kschiffer commented on pkg/webui/console/components/mac-settings-section/validation-schema.js:
```diff
@@ -0,0 +1,124 @@
+        schema.min(0, { ...sharedMessages.validateNumberGte, values: { value: 0 } }).max(7, {
+          ...sharedMessages.validateNumberLte,
+          values: { value: 7 },
+        }),
```
Use `yup.passValues()`

kschiffer commented on pkg/webui/console/components/mac-settings-section/validation-schema.js:
```diff
@@ -0,0 +1,124 @@
+        schema.min(100000, {
+          ...sharedMessages.validateNumberGte,
+          values: { value: 100000 },
+        }),
```
Same.

kschiffer commented on pkg/webui/console/lib/device-utils.js:
```diff
@@ -40,6 +40,12 @@ export const LORAWAN_PHY_VERSIONS = Object.freeze([
   { value: '1.1.0-b', label: 'PHY V1.1 REV B' },
 ])
 
+export const DEVICE_CLASSES = Object.freeze({
```
Nit: Maybe `/console/constants/` is a better place for this?

kschiffer reviewed (changes requested):
Seems like the branch misses a `formik` update to `2.1.5`, I'm getting errors otherwise.

There's also some prop type warnings for the select inputs, which use numbers as values, which is not allowed by the prop types. However, I'd suggest to actually use `number`\-type `<Input />`'s for these anyway. As far as I can see, the values are basically a range of numbers, so using a select for this is an anti-pattern, but I might miss something there (also I feel like this might have been discussed already but I don't remember the conclusion).

kschiffer commented:
Two questions for @rvolosatovs for my understanding:

*   is it possible to change these values after the creation of the end device?
*   Can these options only be set during creation?

rvolosatovs commented:
> Two questions for @rvolosatovs for my understanding:
> 
>     * is it possible to change these values after the creation of the end device?
yes

>     * Can these options only be set during creation?
no

Most of `mac-settings` options, however, are only effective on device MAC state reset - e.g. OTAA (re-)join, ABP FCnt reset, ABP device creation, ABP ResetInd MAC command

Commit:
console: Add mac settings to network server form

Commit:
util: Handle default enum values for mac settings

kschiffer commented on pkg/webui/console/views/device-add/wizard/network-settings-form/index.js:
```diff
@@ -103,10 +126,12 @@ const NetworkSettingsForm = props => {
 
   const validationContext = React.useMemo(
     () => ({
+      isClassB,
       activationMode,
     }),
-    [activationMode],
+    [activationMode, isClassB],
   )
```
I suggest removing `React.useMemo()` here, since this is a really simple function.

kschiffer commented on pkg/webui/console/views/device-general-settings/network-server-form/index.js:
```diff
@@ -59,13 +68,48 @@ const NetworkServerForm = React.memo(props => {
+    activationMode = ACTIVATION_MODES.MULTICAST
+  }
+
   const validationContext = React.useMemo(
```
Same here. I don't think the composition of this context object should be of any performance concern.

kschiffer reviewed (approved):
LGTM now but I'm not very familiar with the specifics of mac settings, so I suggest waiting with merging until @rvolosatovs has greenlighted this as well.

rvolosatovs commented on pkg/webui/components/key-value-map/entry.js:
```diff
@@ -31,43 +31,48 @@ class Entry extends React.Component {
   @bind
   handleRemoveButtonClicked(event) {
     const { onRemoveButtonClick, index } = this.props
+
```
nit; this seems like unnecessary waste of screen real estate

rvolosatovs commented on pkg/webui/console/components/mac-settings-section/index.js:
```diff
@@ -0,0 +1,185 @@
+          max={15}
+        />
+      )}
+      {(isABP || isOTAA) && (
```
All devices are either ABP or OTAA.
This should only be relevant for non-multicast ABP devices

rvolosatovs commented on pkg/webui/console/components/mac-settings-section/index.js:
```diff
@@ -0,0 +1,185 @@
+      initiallyCollapsed={initiallyCollapsed}
+    >
+      <Form.SubTitle title={m.macSettings} />
+      {isABP && (
```
This should only be relevant for non-multicast devices

rvolosatovs replied:
I don't see any change here

rvolosatovs commented on pkg/webui/console/components/mac-settings-section/index.js:
```diff
@@ -0,0 +1,185 @@
+
+  const isClassB = deviceClasses.includes(DEVICE_CLASSES.CLASS_B)
+  const isOTAA = activationMode === ACTIVATION_MODES.OTAA
+  const isABP = activationMode === ACTIVATION_MODES.ABP
```
There should be a distinction here between non-multicast ABP and multicast ABP

rvolosatovs replied:
Still not addressed

rvolosatovs commented on pkg/webui/console/views/device-add/wizard/network-settings-form/index.js:
```diff
@@ -93,6 +97,25 @@ const NetworkSettingsForm = props => {
+  const handleDeviceClassChange = React.useCallback(evt => {
+    const { checked, name } = evt.target
+
+    let cls
```
Exclusion of those 2 characters must have really sped up the development /s
```suggestion
    let class
```

kschiffer replied:
`class` is a keyword in JavaScript.

bafonins replied:
1. yes
2. `class` is reserved in javascript and can be used only to define a class
3. we already use `cls` in many places throughout the codebase

rvolosatovs replied:
If shadowing is not allowed, then that really sucks!
Anyway, just want to point out that this really stands out and is not obvious for a person with not much context about frontend. I wish there was a clearer way to denote this, but I can't think of any.

rvolosatovs commented on pkg/webui/console/views/device-add/wizard/network-settings-form/index.js:
```diff
@@ -93,6 +97,25 @@ const NetworkSettingsForm = props => {
+    const { checked, name } = evt.target
+
+    let cls
+    if (name === 'supports_class_b') {
```
Reverse the order here - we always prefer class C.
Note, that by the spec device can't implement both B and C classes simultaneously.

rvolosatovs commented on pkg/webui/console/views/device-add/wizard/network-settings-form/index.js:
```diff
@@ -93,6 +97,25 @@ const NetworkSettingsForm = props => {
+      cls = DEVICE_CLASSES.CLASS_C
+    }
+
+    if (checked && typeof cls !== 'undefined') {
```
Device can only be in once class at a time, and that is dictated by `mac_state.device_class` and set by Network Server. Console cannot know what class the device is in if either `supports_class_c` or `supports_class_b` is `true`. iff `supports_class_c == supports_class_b == false`, console can be sure that device class is always A.

This whole thing seems too complicated for what it does, the "maximum capable" class(that's the only thing you can determine here, really) is just:
```elm
if supports_class_c then
	DEVICE_CLASSES.CLASS_C
else if supports_class_b then
	DEVICE_CLASSES.CLASS_B
else
	DEVICE_CLASSES.CLASS_A
```

rvolosatovs commented on pkg/webui/console/views/device-add/wizard/network-settings-form/index.js:
```diff
@@ -93,6 +97,25 @@ const NetworkSettingsForm = props => {
+    }
+  }, [])
+
+  const isClassB = deviceClasses.includes(DEVICE_CLASSES.CLASS_B)
```
Isn't there a way to just check the value of `supports_class_b` directly here? Why would `deviceClasses` be necessary at all?

rvolosatovs commented on pkg/webui/console/views/device-add/wizard/network-settings-form/index.js:
```diff
@@ -143,10 +168,17 @@ const NetworkSettingsForm = props => {
+        title={sharedMessages.supportsClassB}
+        name="supports_class_b"
+        component={Checkbox}
+        onChange={handleDeviceClassChange}
```
The device class does not change when this field value is changed. The "maximum capable class" is changed, but the actual class the device is operating in does not change - it's a dynamic value, which may change during the session, which is fully controlled by device.

rvolosatovs commented on pkg/webui/console/views/device-add/wizard/network-settings-form/validation-schema.js:
```diff
@@ -17,21 +17,107 @@ import sharedMessages from '@ttn-lw/lib/shared-messages'
-    }
+    }),
+    rx1_data_rate_offset: Yup.number().when('$activationMode', {
+      is: mode => mode === ACTIVATION_MODES.ABP || mode === ACTIVATION_MODES.OTAA,
```
only non-multicast

rvolosatovs commented on pkg/webui/console/views/device-add/wizard/network-settings-form/validation-schema.js:
```diff
@@ -17,21 +17,107 @@ import sharedMessages from '@ttn-lw/lib/shared-messages'
+      }
+
+      return Yup.object().when('$activationMode', {
+        is: ACTIVATION_MODES.ABP,
```
only non-multicast

rvolosatovs commented on pkg/webui/console/views/device-add/wizard/network-settings-form/validation-schema.js:
```diff
@@ -17,21 +17,107 @@ import sharedMessages from '@ttn-lw/lib/shared-messages'
+      otherwise: schema => schema.strip(),
+    }),
+    resets_f_cnt: Yup.boolean().when('$activationMode', {
+      is: ACTIVATION_MODES.ABP,
```
only non-multicast

rvolosatovs commented on pkg/webui/console/views/device-general-settings/network-server-form/index.js:
```diff
@@ -59,13 +68,48 @@ const NetworkServerForm = React.memo(props => {
     parseLorawanMacVersion(device.lorawan_version),
   )
 
+  const [deviceClasses, setDeviceClasses] = React.useState(() => {
```
I don't think this is necesary

rvolosatovs commented on pkg/webui/console/views/device-general-settings/network-server-form/index.js:
```diff
@@ -59,13 +68,48 @@ const NetworkServerForm = React.memo(props => {
+
+    return deviceClasses
+  })
+  const handleDeviceClassChange = React.useCallback(evt => {
```
IDEM

Commit:
dev: Add base device creation e2e tests

Commit:
console: Fix device class handling in device general settings

Commit:
console,oauth: Add excludePaths prop to wizard form

Allow excluding helper entries from the validation schema

Commit:
console: Fix device class handling in device wizard

bafonins commented:
Blocked by https://github.com/TheThingsNetwork/lorawan-stack/pull/3199

Commit:
dev: Refactor dev:DBRestore target

Make sure that redis is restored as well as cockroach/postgres

bafonins commented:
Blocked by https://github.com/TheThingsNetwork/lorawan-stack/pull/3203

Commit:
console: Make nwk_key required in join settings form

Commit:
console: Fix dev-addr-input for e2e tests

Commit:
console: Fix join-eui-prefixes input

Commit:
util: Fix device parts merge

Allow setting empty object entries to the result object

bafonins commented:
Blocked on https://github.com/TheThingsNetwork/lorawan-stack/pull/3258

rvolosatovs reviewed (changes requested):
Did not review the code as per your explicit request and only checked the UI.
1. It should be possible to unset values. Currently if I set RX1 delay to `4` I see no way to delete the value - I tried saving an empty form, but that did not work.
2. 
![2020-09-30-22:55:39-screenshot](https://user-images.githubusercontent.com/12877905/94738876-5a3b7a80-0370-11eb-9192-57ae3b3d7425.png)
That's not how it works. All devices implement class A. There should be 2 checkboxes here - "Supports class B" and "Supports class C". It is probably fine to make them mutually exclusive, however I'd prefer an approach where both could be selected at the same time. Note, that for multicast devices either "supports class B" or "supports class C" *MUST* be selected
3. I do not like the hidden "advanced settings" approach, in general it's a good idea, but for some cases, e.g. for class B devices in Chinese frequency plan, some of these fields *MUST* be specified and it's confusing for the user to have to search through them. Perhaps you can move the required options out of the hidden menu depending on the device settings?
4. Invalid error handling. NS error is displayed in AS device creation window. As a person unfamiliar with the project I will never figure out what's wrong. What I'd like to see here, is automatic redirection to the place of the error and the field light up in red.
![2020-09-30-23:03:02-screenshot](https://user-images.githubusercontent.com/12877905/94739449-4d6b5680-0371-11eb-9110-f4fa18f9bfcf.png)

rvolosatovs commented:
Also, IMHO, it's better UX to make fields disabled instead of removing them depending on, for example, class selection. That way the user could point at the disabled field and receive a hint regarding why is the field disabled.

bafonins commented:
@rvolosatovs 

>> It should be possible to unset values. Currently if I set RX1 delay to 4 I see no way to delete the value - I tried saving an empty form, but that did not work.

fixed

>> That's not how it works. All devices implement class A. There should be 2 checkboxes here - "Supports class B" and "Supports class C". It is probably fine to make them mutually exclusive, however I'd prefer an approach where both could be selected at the same time. Note, that for multicast devices either "supports class B" or "supports class C" MUST be selected

fixed

>> Invalid error handling. NS error is displayed in AS device creation window. As a person unfamiliar with the project I will never figure out what's wrong. What I'd like to see here, is automatic redirection to the place of the error and the field light up in red.

Will implement this in a separate PR. https://github.com/TheThingsNetwork/lorawan-stack/issues/3291

>>  I do not like the hidden "advanced settings" approach, in general it's a good idea, but for some cases, e.g. for class B devices in Chinese frequency plan, some of these fields MUST be specified and it's confusing for the user to have to search through them. Perhaps you can move the required options out of the hidden menu depending on the device settings?

Now we expand the advanced settings section by default for multicast devices and make the `ping_slot_periodicity` required. Are there any other MAC settings required for other device configurations?

rvolosatovs reviewed (changes requested):
![2020-10-07-21:17:02-screenshot](https://user-images.githubusercontent.com/12877905/95377179-801dcd80-08e2-11eb-86f7-eb0870419082.png)
NetID or Net ID (@johanstokking ?), KEK labels

I still do not see the the ping slot frequency field and have to manually expand "advanced settings". That's a recipe for disaster, especially considering that NS errors are represented as JS errors (in the screenshot you get a false impression of the UX, because the request is only sent from JS setting tab and that's the selector, which lights up red)
![2020-10-07-21:20:00-screenshot](https://user-images.githubusercontent.com/12877905/95377459-e4d92800-08e2-11eb-81ce-7a014641e743.png)

rvolosatovs commented:
Additionally, after a creation error happens, when I go back to "Network layer settings", the "Supports Class B" checkbox is unchecked, even though it was checked before by me. That also should not happen

rvolosatovs commented:
Additionally, the user should be able to change the device type and MAC version by just going back through the tabs. Currently to do that, you have to restart the process, which also should not happen

rvolosatovs commented:
![2020-10-07-21:25:59-screenshot](https://user-images.githubusercontent.com/12877905/95378084-b60f8180-08e3-11eb-9e40-b83f2254c85e.png)
If no ID is set (i.e. the field is not selected) and "Network layer settings" button is pressed - nothing happens, the error also does not appear. The error only starts appearing after something is typed in the field (and, for example, cleared after)

rvolosatovs commented:
![2020-10-07-21:29:16-screenshot](https://user-images.githubusercontent.com/12877905/95378417-2cac7f00-08e4-11eb-94f9-5436fa37dd54.png)
Again, not a fan of this, IMO these should just be 2 checkboxes without title. Otherwise, the title could be "LoRaWAN class capabilities"

johanstokking commented:
> NetID or Net ID (@johanstokking ?)

NetID

Commit:
console: Fix wizard form values reset

Commit:
console: Improve navigation through steps in device wizard

Commit:
console: Move bit_f_cnt to mac-settings section

Commit:
console: Expand advanced settings in ns form on error

Commit:
dev: Update locales

Commit:
console: Fix basic-settings form device id validation

bafonins commented:
@rvolosatovs 

> It should be possible to unset values. Currently if I set RX1 delay to 4 I see no way to delete the value - I tried saving an empty form, but that did not work.

fixed

> Invalid error handling. NS error is displayed in AS device creation window. As a person unfamiliar with the project I will never figure out what's wrong. What I'd like to see here, is automatic redirection to the place of the error and the field light up in red.

fixed. Now the user is directed to the errored step.

> If no ID is set (i.e. the field is not selected) and "Network layer settings" button is pressed - nothing happens, the error also does not appear. The error only starts appearing after something is typed in the field (and, for example, cleared after)

fixed

> Again, not a fan of this, IMO these should just be 2 checkboxes without title. Otherwise, the title could be "LoRaWAN class capabilities"

changed `End device class` to `LoRaWAN class capabilities`

> NetID, KEK labels

fixed

> I still do not see the the ping slot frequency field and have to manually expand "advanced settings". That's a recipe for disaster, especially considering that NS errors are represented as JS errors

With this PR the user is redirected to the errored step, also I added functionality to expand the advanced settings section on errored submission.

rvolosatovs commented:
![2020-11-19-19:05:43-screenshot](https://user-images.githubusercontent.com/12877905/99705616-407a0200-2a9a-11eb-987a-3acc2f163a96.png)
When creating an OTAA device
1. `invalid field mask` is completely opaque to the user
2. There is no way to set the required class B parameter at all

Please parse the error, and make sure the wrong field is somehow highlighted

bafonins commented:
> There is no way to set the required class B parameter at all

Fixed this. No the NS form properly shows class b related fields on error. Now the user is able to recover from not filling in the `ping_slot_frequency` value.

Wont add this functionality in this PR.

bafonins commented:
@rvolosatovs

bafonins commented:
@rvolosatovs

bafonins commented:
@rvolosatovs

rvolosatovs reviewed (approved):
Please avoid unnecessary spam notifications in the future - I was off for a few days, you can see that in the company calendar

Commit:
util: Format end device entity map

Commit:
util: Add skip_payload_crypto to device-entity-map
